File: NNN-Solution/NNN/Components/Autodiff/TensorGraph.cs
namespace NNN.Components.Autodiff;

public partial class Tensor
{
    // Autograd graph functions

    /// <summary>
    /// Creates a new gradient array which matches the size of the linear value array of the tensor.
    /// </summary>
    public void RestoreGrad()
    {
        if (RequiresGrad) Grad = new double[ElementCount];
    }

    /// <summary>
    /// Clears the parent and gradient function data of the tensor.
    /// </summary>
    public void ClearGraph()
    {
        _parents.Clear();
        _backward = delegate { };
    }

    /// <summary>
    /// Increments the current autgrad graph generation.
    /// </summary>
    public static void BeginForward()
    {
        if (!Inference) _forwardGen++;
    }

    /// <summary>
    /// Prepares the autograd graph node for another forward pass.
    /// </summary>
    void PrepareForward()
    {
        if (_lastGen == _forwardGen) return;

        _opIndex = 0;

        foreach (var r in _results)
        {
            r.ClearGraph();
        }

        _lastGen = _forwardGen;
    }

    /// <summary>
    /// Finalizes the autograd graph node's internal data after a forward pass.
    /// </summary>
    void FinalizeForward()
    {
        // Trim any excess result tensor references not used in the last autograd graph
        if (_opIndex < _results.Count)
        {
            _results.RemoveRange(_opIndex, _results.Count - _opIndex);
        }
    }

    /// <summary>
    /// Calculates the gradients of all tensor nodes in the current autograd graph.
    /// </summary>
    public void Backward()
    {
        // Initialize topography and visited node buffers if not yet initialized
        _topo ??= [];
        _visited ??= [];

        // Clear previous topography and visited nodes
        _topo.Clear();
        _visited.Clear();

        BuildTopo(this, _topo, _visited); // build topography of current graph

        // Zero out the gradients of all nodes
        foreach (var t in _topo)
        {
            if (t.RequiresGrad) Array.Clear(t.Grad, 0, t.GradCount);
        }

        Array.Fill(Grad, 1.0); // initialize current node's gradient to 1 (assume Backward() was called on the final node)

        // Iterate backwards from last node in the graph
        for (int i = _topo.Count - 1; i >= 0; i--)
        {
            _topo[i]._backward();
        }

        // Finalize forward pass for each node
        foreach (var t in _topo)
        {
            t.FinalizeForward();
        }
    }

    /// <summary>
    /// Builds the topography of the current autograd graph.
    /// </summary>
    /// <param name="t">Node which is being added to the topography.</param>
    /// <param name="topo">Buffer to store topography in.</param>
    /// <param name="visited">HashSet to track visited nodes in.</param>
    static void BuildTopo(Tensor t, List<Tensor> topo, HashSet<Tensor> visited)
    {
        if (visited.Contains(t)) return; // skip if node appears in graph multiple times
        visited.Add(t);

        // Recursively add all parent nodes to the topography
        foreach (var p in t._parents)
        {
            BuildTopo(p, topo, visited);
        }

        topo.Add(t); // append current node at the end of the topography
    }
}

File: NNN-Solution/NNN/TensorGraph.cpp
#include "pch.h"
#include "Tensor.h"

/* Autograd graph functionality */

// Restores and resets the gradient vector of the tensor.
void Tensor::restore_grad()
{
	if (requires_grad)
	{
		_grad.assign(element_count(), 0.0);
	}
}

void Tensor::clear_graph()
{
	_parents.clear();
	_backward = [] {};
}

void Tensor::begin_forward()
{
	_forward_gen++;
}

void Tensor::prepare_forward()
{
	// Clear autograd graph connections and reset operation index if not yet participated in the current forward pass

	if (_last_gen == _forward_gen) return;

	_op_index = 0;

	for (std::shared_ptr<Tensor>& t : _results)
	{
		t->clear_graph();
	}

	_last_gen = _forward_gen;
}

void Tensor::finalize_forward()
{
	// Prune any unused result tensor allocations - prevent potential memory leaks
	if (_op_index < (int)_results.size())
	{
		_results.resize(_op_index);
	}
}

void Tensor::build_topo(const std::shared_ptr<Tensor>& t, std::vector<std::shared_ptr<Tensor>>& topo,
	std::unordered_set<std::shared_ptr<Tensor>, TensorPtrHash, TensorPtrEqual>& visited)
{
	// Skip tensor if already visited in topography
	if (visited.contains(t)) return;
	visited.insert(t);

	// Recursively build topography starting from this tensor
	for (std::shared_ptr<Tensor>& p : t->_parents)
	{
		build_topo(p, topo, visited);
	}

	topo.push_back(t);
}

void Tensor::backward()
{
	// Initialize _topo and _visited if not yet used
	if (!_topo.has_value()) _topo.emplace();

	if (!_visited.has_value()) _visited.emplace();

	auto& topo = *_topo;
	build_topo(shared_from_this(), topo, *_visited);
	_visited->clear();

	// Zero out previous gradients
	for (std::shared_ptr<Tensor>& t : topo)
	{
		t->restore_grad();
	}

	// Set initial gradient
	_grad.assign(_grad.size(), 1.0);

	// Compute upstream gradient from each node in topography
	for (int i = (int)topo.size() - 1; i >= 0; i--)
	{
		topo[i]->_backward();
	}

	// Prune unused result tensor allocations
	for (std::shared_ptr<Tensor>& t : topo)
	{
		t->finalize_forward();
	}

	_topo->clear();
}